pkg/lib/bundle/validate.go

```diff
@@ -0,0 +1,256 @@
+				case strings.TrimSuffix(MetadataDir, "/"):
+					log.Info("Found metadata directory")
+					metadataFound = true
+					annotationsDir = filepath.Join(dir, MetadataDir)
+				}
+			}
+		}
+	}
+
+	if manifestsFound == false {
+		return fmt.Errorf("Unable to locate manifests directory")
+	} else if metadataFound == false {
+		return fmt.Errorf("Unable to locate metadata directory")
+	}
+
+	log.Info("Getting mediaType info from manifests directory")
+	mediaType, err := GetMediaType(manifestsDir)
+	if err != nil {
+		log.Error("Unable to determine manifests mediaType")
+		return err
+	}
+
+	// Validate annotations.yaml
+	file, err = ioutil.ReadFile(filepath.Join(annotationsDir, AnnotationsFile))
+	if err != nil {
+		log.Error("Unable to validate annotations.yaml")
+		return err
+	} else {
+		if err = ValidateBundleAnnotations(mediaType, file); err != nil {
+			return err
+		}
+	}
+
+	log.Info("All validation tests have been completed successfully")
+
+	return nil
+}
+
+func UntarFile(dir, tarName string) error {
+	file, err := os.Open(filepath.Join(dir, tarName))
+	if err != nil {
+		return err
+	}
+
+	err = Untar(file, dir)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func PullImage(imageTag, imageBuilder string) (*exec.Cmd, error) {
+	var args []string
+
+	switch imageBuilder {
+	case "docker", "podman":
+		args = append(args, "pull", imageTag)
+	default:
+		return nil, fmt.Errorf("%s is not supported image builder", imageBuilder)
+	}
+
+	return exec.Command(imageBuilder, args...), nil
+}
+
+func SaveImage(directory, imageTag, imageBuilder string) (*exec.Cmd, error) {
+	var args []string
+
+	switch imageBuilder {
+	case "docker", "podman":
+		args = append(args, "save", "-o", filepath.Join(directory, BundleTarFile), imageTag)
+	default:
+		return nil, fmt.Errorf("%s is not supported image builder", imageBuilder)
+	}
+
+	return exec.Command(imageBuilder, args...), nil
+}
+
+func ParseManifestJson(file []byte) (string, []string, error) {
+	var manifest = []*ManifestJson{}
+
+	err := json.Unmarshal(file, &manifest)
+	if err != nil {
+		log.Errorf("Unable to parse manifest.json")
+		return "", nil, err
+	}
+
+	return manifest[0].Config, manifest[0].Layers, nil
+}
+
+func ValidateBundleAnnotations(mediaType string, file []byte) error {
+	var fileAnnotations AnnotationMetadata
+	var invalid bool
+
+	annotations := map[string]string{
+		MediatypeLabel:      mediaType,
+		ManifestsLabel:      ManifestsDir,
+		MetadataLabel:       MetadataDir,
+		PackageLabel:        "",
+		ChannelsLabel:       "",
+		ChannelDefaultLabel: "",
+	}
+
+	log.Info("Validating annotations.yaml")
+
+	err := yaml.Unmarshal(file, &fileAnnotations)
+	if err != nil {
+		log.Error("Unable to parse annotations.yaml")
+		return err
+	}
+
+	for label, item := range annotations {
+		val, ok := fileAnnotations.Annotations[label]
+		if ok {
+			log.Infof(`Found annotation "%s" with value "%s"`, label, val)
+		} else {
+			log.Errorf(`Missing annotation "%s"`, label)
```

Comment:
Rather than logging the validation errors, would it be reasonable to define a `ValidationError` struct? Something like:
```go
type ValidationError struct {
    AnnotationErrors []error
}

// Error ensures ValidationError implements the error interface
func (v ValidationError) Error() string { ... }
```

Then we could append each of the annotation errors to the slice as we're iterating. This type of design would give callers more flexibility on how to handle each of the errors. For example:

```go
 if err := bundle.ValidateBundleAnnotations(mediaType, data); err != nil {
    var verr *bundle.ValidationError
    if errors.As(err, &verr) {
        for _, ae := range verr.AnnotationErrors {
            // handle each annotation error
        }
    } else {
        // got an error not related to validation
        return err
    }
}
```